### PR TITLE
Notebook expansion

### PR DESCRIPTION
**Pull request by giordanoscerra**
Final review, everyone gather!

### CODE REVIEW

File: Notebook.ipynb [hunk 1]
```diff
@@ -44,22 +44,23 @@
...
+    "- If it deems so, fights monsters and hunger\n"
    ]
   },
   {
    "cell_type": "markdown",
    "id": "9d7dd98e",
    "metadata": {},
    "source": [
-    "The one listed are not only steps that are necessary to perform the task successfully, but also *subtasks* on their own. As it will be thoroughly described later, the task is solved by selecting the most appropriate subtask and execute it. The choice of the most appropriate subtask is controlled by a knowledge base, while a python class is in charge of the proper execution."
+    "Those listed here are not only necessary steps to perform the task successfully, but also *subtasks* on their own. As it will be thoroughly described later, the task is solved by continuously selecting the most appropriate subtask and executing it, given the state of the environment at a given moment. The execution of the subtask is controlled by a knowledge base which is updated it with the changes that are applied to the environment."
```
Comment (giordanoscerra) [inline Notebook.ipynb:63]: *which is updated with

File: Notebook.ipynb [hunk 2]
```diff
@@ -136,47 +139,66 @@
    "id": "0b23fee1",
    "metadata": {},
    "source": [
-    "As discussed during the oral presentation, and as described in the previous paragraph, the knowledge base is where both the information is stored and the decision making process for the agent takes place. This entity itself is divided in two parts, which can be distinguished by the two different languages in which they're written: Prolog and Python. In fact, Python is used as a wrapper interface to control the Prolog part, thanks to the open source library pyswip. \n",
+    "As discussed during the oral presentation, and as described in the previous paragraph, the knowledge base is where both the information is stored and the decision making process for the agent takes place. This entity itself is divided in two parts, which can be distinguished by the two different languages in which they're written: Prolog and Python. In fact, Python is used as a wrapper interface to control the Prolog part, thanks to the open source library pyswip.\n",
```
Comment (giordanoscerra) [inline Notebook.ipynb:142]: ~"As discussed during the oral presentation, and as described in the previous paragraph"~

File: Notebook.ipynb [hunk 3]
```diff
@@ -136,47 +139,66 @@
...
     "The subtasks in the knowledge base are defined in order of relevance. We have:\n",
     "- getCarrot\n",
     "- feedSteed\n",
     "- getSaddle\n",
     "- applySaddle\n",
-    "- explore (this subtask is chosen when no premise is satisfied for another subtask to start)\n",
+    "- explore (chosen when no premise is satisfied for another subtask to start)\n",
     "- eat\n",
     "- attackEnemy\n",
     "\n",
```
Comment (giordanoscerra) [inline Notebook.ipynb:151]: this is ..wrong? eat and attackEnemy are the firsts!

File: Notebook.ipynb [hunk 4]
```diff
@@ -136,47 +139,66 @@
...
+    "In the (rare) case where the agent has no better action than 'explore' and it performs 1 full map patrol, the subtask switches to riding. This was implemented to avoid endless exploring, it's a desperate last resort option."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "2fa569b7",
+   "metadata": {},
+   "source": [
+    "The first subtask that is listed in the knowledge base is `action(attackEnemy)`, which can be entailed if `attack(enemy,_)` can be entailed. This counterintuitive syntax is a workaround to maintain the compatibility with the way that the `Agent.act()` method selects the other subtasks while maintaining the possibility to query which enemy to attack. This information is necessary to execute the *attackEnemy* subtask, and is better provided by the knowledge base; on the other hand the other subtasks don't have arguments tied to logic variables of any sort, since they don't need them: all the relevant information such as which is the closest element to reach is more easily provided by processing through python the information present in the knowledge base or other data structures."
```
Comment (giordanoscerra) [inline Notebook.ipynb:163]: attackEnemy is the only clause explained and it should be the least relevant to the task ... ? i'd remove or shorten this part (saying briefly we do this for surviving purposes)

File: Notebook.ipynb [hunk 5]
```diff
@@ -136,47 +139,66 @@
...
-    "*attackEnemy* is the first that is listed as we wanted it to be at higher priority, in the case that it could be inferred together with some other subtask. For the task at hand, the subtasks can be totally ordered (ordered in the sense of preference) if no complications arise: the ideal order should be *getCarrot*, *feedSteed*, *getSaddle*, *applySaddle*, *rideSteed*. The *attackEnemy* and *eat* subtasks are sort of contingency plans, *explore* is a last resort. It is natural that the first two are prioritize above anything else if there is an interruption and that the latter has least priority."
+    "At a higher level, almost all the subtasks are performed in the same way: go to the closer specified element and interact with it by applying an action. The latter part is managed by the `Agent.perform_action()` method, while for the former part we used A*. This algorithm is well suited for targets that are static, but sometimes the agent has to go towards elements that change their position in the map (e.g. moving enemies, ponies). This problem was resolved by considering only the first step of the entire A* path, reiterated repeatedly according to the new target position."
```
Comment (giordanoscerra) [inline Notebook.ipynb:189]: "reiterated repeatedly" , i wrote this and i hate it

File: Notebook.ipynb [hunk 6]
```diff
@@ -194,7 +216,8 @@
    "source": [
     "Let's see a small demo of our project. In order to run the following code, and in general to replicate our work, make sure that:\n",
     "1. this notebook stays inside the folder in which it was delivered. This is necessary to resolve the imports\n",
-    "2. the following packages that were used in the hands-on sessions are installed: `pyswip`, `minihack`, `numpy`, `nle`, `matplotlib`, `notebook` "
+    "2. the following packages that were used in the hands-on sessions are installed: `pyswip`, `minihack`, `numpy`, `nle`, `matplotlib`, `notebook` \n",
+    "3. the `pandas` package is installed, in order to plot the data in the assessment paragraph"
```
Comment (giordanoscerra) [inline Notebook.ipynb:220]: May be needed to install another package for dealing with excels ..
i don't remember how it's called tho ... hehe

File: Notebook.ipynb [hunk 7]
```diff
@@ -194,7 +216,8 @@
...
    ]
   },
   {
```
Comment (giordanoscerra) [inline Notebook.ipynb:223]: are we sure euclidean distance works..?

File: Notebook.ipynb [hunk 8]
```diff
@@ -295,15 +318,15 @@
    "outputs": [],
    "source": [
     "while(not level.is_episode_over()):\n",
-    "    steinbeck.act(level, show_steps=True, graphic=True, delay=0.1)\n"
+    "    steinbeck.act(level, show_steps=True, graphic=True, delay=0.1)"
```
Comment (giordanoscerra) [inline Notebook.ipynb:321]: yep this is a long abb movie ... and the messages are quite unreadable...

File: Notebook.ipynb [hunk 9]
```diff
@@ -295,15 +318,15 @@
...
    ]
   },
   {
    "cell_type": "markdown",
    "id": "c3b90bdb",
    "metadata": {},
    "source": [
-    "Let's see how well our valiant Steinbeck performed in this level"
+    "Let's see how well our agent performed in this level"
```
Comment (giordanoscerra) [inline Notebook.ipynb:329]: this just returned me 999  ????????????

File: Notebook.ipynb [hunk 10]
```diff
@@ -321,7 +344,14 @@
...
+    "\n",
+    "First of all, why a sum? \n",
+    "Since we decided to append the reward assigned in each state of the environment to a list, we end up with a pretty long one, especially in gargantuan mazes.\n",
+    "The events we defined and chose to record are `getCarrots` and `MountEvent` which respectively assign a reward of one point for picking up a carrot and 1000 points for mounting the steed.\n",
+    "\n",
+    "We discussed extensively about using an event that'd give off a reward of -1 for each step taken by the agent and came to the conclusion that it would be inadequate for our task: we wished to face multiple contingencies while taming the steed: fighting a monster in our path to a carrot or near the steed, finding all the carrots, getting the steed to drop the saddle if it ever wished to take it and many more. Because of this, using such a reward would have pushed us to solve the task *faster* instead of *better*, becoming a sort of toxic inductive bias on our work: we'd have been more focused on finding good heuristics to finish the level in time instead of solving the many problems we had to face. \n",
```
Comment (giordanoscerra) [inline Notebook.ipynb:353]: remove double ":"

Comment (giordanoscerra) [reply Notebook.ipynb:353]: update, this is full of ":" . why !!!!!!!!!!!

File: Notebook.ipynb [hunk 11]
```diff
@@ -334,22 +364,88 @@
...
+    "The general behaviour of the agent has been tested over the levels 0, 1, 2, 3 and -1. The latter is a level where the agent is randomly put in a randomly generated maze, level 0 represents a small, rectangular room, level 1 is composed by two different rooms, connected by a narrow corridor, level 2 is a level that more closely resembles a NetHack level, and finally level 3 which is a fairly complex maze with narrow corridors.\\\n",
+    "Furthermore, we evaluated the behavior of our agent under varying degrees of problem relaxations by testing it both with and without the contingencies presented by hunger and monsters.\n",
+    "\n",
+    "By running the cell below with the appropriate parameters it is possible to see two histograms, one indicating the medium steps on a success and failure scenario, and the other indicating the success rate of our agent in a certain level with certain conditions.\n",
+    "The naming convention we devised for the different evaluation environments is given by the concatenation of these elements:\n",
+    "- L\n",
```
Comment (giordanoscerra) [inline Notebook.ipynb:376]: "L (as for level)"

File: Notebook.ipynb [hunk 12]
```diff
@@ -334,22 +364,88 @@
...
+    "- a number between -1 and 3 to indicate the level\n",
+    "- Y or N to denote either the presence or the absence of a certain contingency \n",
+    "- A, which is the initial of \"Apples\" (eaten by our agent): denotes the contingency of hunger \n",
+    "- Y or N to indicate either the presence or the absence of a certain contingency\n",
```
Comment (giordanoscerra) [inline Notebook.ipynb:380]: "Y or N (as above)"

File: Notebook.ipynb [hunk 13]
```diff
@@ -334,22 +364,88 @@
...
+    "\n",
+    "show_stats(level=2, success_rate_flag=True)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "ead24e7f",
+   "metadata": {},
+   "source": [
+    "It can be immediately noticed that the success rate for level 1 is very low if there are enemies. This is due to the map's shape, that often makes it impossible to be solved. In some lucky cases (namely, if a carrots spawns very close to the agent) the task can be completed in the level.\n",
```
Comment (giordanoscerra) [inline Notebook.ipynb:409]: we first talk about level 1, remember to set level 1 in the python cell above

Comment (giordanoscerra) [reply Notebook.ipynb:409]: typo: "a carrot" instead of carrots

File: Notebook.ipynb [hunk 14]
```diff
@@ -334,22 +364,88 @@
...
+    "\n",
+    "We expected that the rates of success would decrease as the dimension of the levels grew, in fact the number of steps that the agent had to take to complete the task increased to a point in which he would sometimes die of starvation. This is what led us to add the *Eat* subtask. We were surprised to find out that starvation isn't very impactful on the success rate: on levels 0,1 and 2 the difference in success rates is negligible, in levels -1 and 3 the number of performed steps in case of failure increases drastically but the success rate barely changes. We can say that in the latter case the agony of our agent Steinbeck is prolonged.\n",
```
Comment (giordanoscerra) [inline Notebook.ipynb:411]: typo: "has" instead of "had" ?

File: Notebook.ipynb [hunk 15]
```diff
@@ -334,22 +364,88 @@
...
+    "\n",
+    "Finally, we can notice that when the agent succeeds the number of steps remains fairly constant in a level. That is, this number is almost unaffected by the contingencies and is only tied to the map's complexity. Given that the success rate is usually very high, that is an indication of the fact that the contingencies are managed very effectively by our agent Steinbeck. This conclusion can be drawn also by direcly observing the execution of the code, not just by observing the data.\n",
```
Comment (giordanoscerra) [inline Notebook.ipynb:415]: add comma after succeeds

Comment (giordanoscerra) [reply Notebook.ipynb:415]: typo: directly instead of direcly

File: Notebook.ipynb [hunk 16]
```diff
@@ -334,22 +364,88 @@
...
+   "source": [
+    "from stats.show_stats import show_stats\n",
+    "\n",
+    "# -1 for random maze\n",
+    "# 0 for square\n",
+    "# 1 for impossible one\n",
+    "# 2 for nethack-like level\n",
+    "# 3 for complex maze\n",
+    "\n",
+    "# success_rate_flag must be set to True to see the rate of success of the agent\n",
```
Comment (giordanoscerra) [inline Notebook.ipynb:399]: i would talk somewhere around here that these numbers come from a 200-2000 times executions

File: Notebook.ipynb [hunk 17]
```diff
@@ -360,6 +456,23 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "2560cec6",
+   "metadata": {},
+   "source": [
+    "Despite adding challenges such as enemies and hunger, the success rate remains high in most of the maps, many of which are very complex environments. The approach used in this project to tackle this task revealed to be both very effective and scalable.\\ \n",
+    "The project could be extended in many ways, such as adding more challenging enemies, using different algorithms to approach dynamic elements in the map, adding traps and many others. In general, this project presents many further developements that could be considered by whoever wants to tackle this interesting task. We could very well say, quoting Nobel prize John Steinbeck: \"*Many trips continue long after movement in time and space have ceased.*\""
```
Comment (giordanoscerra) [inline Notebook.ipynb:473]: remove \ after scalable

**Review by giordanoscerra — COMMENTED** [review]
# m y  w r i t i n g  t i p s 

## don't feel free to ignore

File: Notebook.ipynb (hunk 7 above)
Comment (M4rinz) [reply Notebook.ipynb:223]: Euclidean distance only needs NumPy, it should work with only that package

File: Notebook.ipynb (hunk 8 above)
Comment (M4rinz) [reply Notebook.ipynb:321]: It is what it is.
Life is life. 
Stay with me

File: Notebook.ipynb (hunk 10 above)
Comment (M4rinz) [reply Notebook.ipynb:353]: replaced second : with , such as

File: Notebook.ipynb (hunk 11 above)
Comment (M4rinz) [reply Notebook.ipynb:376]: ok

File: Notebook.ipynb (hunk 14 above)
Comment (M4rinz) [reply Notebook.ipynb:411]: Corrected to a hypothetical sentence

**Pull request merged**